src/chain_sync/chain_muxer.rs

```diff
@@ -365,28 +370,48 @@ where
                 metrics::PEER_TIPSET_EPOCH
                     .with_label_values(&[source.to_string().as_str()])
                     .set(request.heaviest_tipset_height);
+                if let Ok(Some(tipset)) = Tipset::load(
+                    chain_store.blockstore(),
+                    &TipsetKeys::from_iter(request.heaviest_tip_set),
+                ) {
+                    network
+                        .peer_manager()
+                        .update_peer_head(source, tipset.into())
+                        .await;
+                }
+
                 return Ok(None);
             }
             NetworkEvent::HelloResponseOutbound { request, source } => {
                 metrics::LIBP2P_MESSAGE_TOTAL
                     .with_label_values(&[metrics::values::HELLO_RESPONSE_OUTBOUND])
                     .inc();
                 let tipset_keys = TipsetKeys::from_iter(request.heaviest_tip_set);
-                let tipset = match Self::get_full_tipset(
-                    network.clone(),
-                    chain_store.clone(),
-                    source,
-                    tipset_keys,
-                )
-                .await
-                {
-                    Ok(tipset) => tipset,
-                    Err(why) => {
-                        debug!("Querying full tipset failed: {}", why);
-                        return Err(why);
+                // Retrieve tipset from store based on passed in TipsetKeys
+                if let Ok(Some(tipset)) = chain_store.load_tipset(&tipset_keys) {
+                    // Skip old tipsets
+                    if chain_store.is_epoch_finalized(tipset.epoch()) {
+                        return Ok(None);
+                    } else {
+                        let full_tipset = match Self::get_full_tipset(
+                            network.clone(),
+                            chain_store.clone(),
+                            source,
+                            &tipset,
+                        )
+                        .await
+                        {
+                            Ok(tipset) => tipset,
+                            Err(why) => {
+                                debug!("Querying full tipset failed: {}", why);
+                                return Err(why);
+                            }
+                        };
+                        (full_tipset, source)
```

Comment:
nit: if you return here - you can get rid of the `else` block below and just do a return there too.

Reply:
The `full_tipset` is assigned to a variable and processed later, early return here breaks the logic.